Breaking up the Rational Factors

AnimeshSinha1309 opened the pull request:
`Factors()` splits rational factors into its numerator and denominator.
`collect_const()` can now collect the common dividing factors in an expression.
 
#### References to other Issues or PRs
Fixes #16296 

#### Brief description of what is fixed or changed
Breaks down the numerator and denominator of Rational factors in a multiplicative expression, fixing the denominator collection issue that collect_const() was having.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * sympy.core.exprtools.Factors() now splits each rational factor into a seaprate numerator and denominator
<!-- END RELEASE NOTES -->

AnimeshSinha1309 committed:
Breaking up the Rational Factors

Fixes #16296. Breaks down the numerator and denominator of Rational factors in a multiplicative expression, fixing the denominator collection issue that collect_const() was having.

jksuom commented on sympy/core/exprtools.py:
```diff
@@ -355,6 +354,12 @@ def __init__(self, factors=None):  # Factors
+            # Handle all rational Coefficients
+            for f in list(factors.keys()):
+                if type(f) is Rational or type(f) is Half:
+                    factors[f.p] = (factors[f.p] if f.p in factors else 0) + factors[f]
```
Perhaps `factors.set_default(f.p, 0)` could be used.

jksuom replied:
`f.p` is a Python `int`. It should be converted into a SymPy `Integer`.

jksuom commented on sympy/core/exprtools.py:
```diff
@@ -355,6 +354,12 @@ def __init__(self, factors=None):  # Factors
                 c.remove(I)
             factors = dict(Mul._from_args(c).as_powers_dict())
+            # Handle all rational Coefficients
+            for f in list(factors.keys()):
```
Would it suffice to write `for f in factors.keys():`?

AnimeshSinha1309 replied:
No, the `factors.keys()` method returns an iterator and as I change the dictionary the iterator throws an error.

oscarbenjamin commented on sympy/core/exprtools.py:
```diff
@@ -355,6 +354,12 @@ def __init__(self, factors=None):  # Factors
             factors = dict(Mul._from_args(c).as_powers_dict())
+            # Handle all rational Coefficients
+            for f in list(factors.keys()):
+                if type(f) is Rational or type(f) is Half:
```
This could be `if isinstance(f, Rational)`

oscarbenjamin replied:
And then `Half won't need importing above.

AnimeshSinha1309 replied:
Did both of those things.

oscarbenjamin commented:
The tests seem to be breaking right now so I guess something here isn't working right.

This should have some new tests when it is working.

AnimeshSinha1309 committed:
Fixing the issues

Converting p, q to Sympy Integer and checking instance of Rational to fix the errors in tests.

jksuom commented on sympy/core/exprtools.py:
```diff
@@ -356,9 +356,10 @@ def __init__(self, factors=None):  # Factors
+                    p, q = Integer(f.p), Integer(f.q)
+                    factors[p] = (factors[p] if p in factors else 0) + factors[f]
+                    factors[q] = (factors[q] if q in factors else 0) - factors[f]
                     factors.pop(f)
```
If `f` is integer, then `p` is `f` (and `q` is 1, which is not useful). This statement will delete `factors[p]`. The condition should probably be `isinstance(f, Rational) and not isinstance(f, Integer)`.

AnimeshSinha1309 replied:
Thank you so much, got everything fixed and working.

AnimeshSinha1309 committed:
Not breaking up Integers

All tests working now. Changed condition to Rational and Not integer when breaking up numberator and denominator.

jksuom commented:
Please add a unit test.

AnimeshSinha1309 committed:
Added a couple of tests

Testing factors directly and checking that collect_const can collect rationals.

jksuom commented on sympy/simplify/tests/test_radsimp.py:
```diff
@@ -342,6 +342,9 @@ def test_collect_const():
         2*sqrt(sqrt(2) + 3)*(sqrt(5)*x + y) + 2
 
+    # issue 16296
+    assert collect_const(a + b + x/2 + y/2) != a + b + (x/2 + y/2)
```
Perhaps this could be `== a + b + Mul(S.Half, x + y, evaluate=False)` (though it does not look so nice).

AnimeshSinha1309 replied:
Yeah, I did not know how to stop evaluation, thanks.

AnimeshSinha1309 committed:
Improving one test

Not evaulating Mul allows testing the collect_const function with Rational Factors.

AnimeshSinha1309 commented:
Hope everything works well now, thanks a lot.

jksuom commented on sympy/core/exprtools.py:
```diff
@@ -9,7 +9,7 @@
 from sympy.core.expr import Expr
 from sympy.core.sympify import sympify
-from sympy.core.numbers import Rational, Integer, Number, I
+from sympy.core.numbers import Rational, Integer, Number, I, Half
```
The class `Half` is no more needed.

AnimeshSinha1309 replied:
Done with it.

AnimeshSinha1309 committed:
Removed redundant import

Importing sympy.core.numbers.Half is no longer necessary.

jksuom commented:
Thanks, looks good.
